### PR TITLE
fix error when add new service with two same backends

### PR DESCRIPTION
**Pull request by javu7762**
fix error when add new service with two same backends

### As-Is: can't delete a service in lb-only mode.
   After create a service with two backend with same ip address and port, 
   deleting that service fails because serviceKey of bpf map cilium_lb4_services_v2  with  **slot 2 backend**  does not exist.
   Also  after pod restarts and restores service from bpf,  service have only one backend

```
/home/cilium# cilium service update  --id=101  --frontend=3.3.3.3:80 --backends=1.1.1.1:80,1.1.1.1:80    --k8s-external
Creating new service with id '101'
Added service with 2 backends
/home/cilium# cilium service list
ID    Frontend     Service Type   Backend
101   3.3.3.3:80   ExternalIPs    1 => 1.1.1.1:80 (active)
                                  2 => 1.1.1.1:80 (active)
/home/cilium# cilium service delete 101
Error: [DELETE /service/{id}][500] deleteServiceIdFailure  Unable to delete service entry 3.3.3.3:20480: unable to delete element 3.3.3.3:80 from map cilium_lb4_services_v2: no such file or directory
```

> Error: [DELETE /service/{id}][500] deleteServiceIdFailure  Unable to delete service entry 3.3.3.3:20480: slot 2, unable to delete element 3.3.3.3:80 from map cilium_lb4_services_v2: no such file or directory
   
   - when pod is restarting and restoring from bpf map without deleting service, a backend disappears.

```
/home/cilium#  cilium service list
ID    Frontend     Service Type   Backend
101   3.3.3.3:80   ExternalIPs    1 => 1.1.1.1:80 (active)
```   

### To-Be
   - disallow creating a service with same backends with same ip address and port

```
# cilium service update  --id=101  --frontend=3.3.3.3:80 --backends=1.1.1.1:80,1.1.1.1:80    --k8s-external
Creating new service with id '101'
Error: Cannot add/update service: [PUT /service/{id}][500] putServiceIdFailure  duplicated backends are not allowed
```

### CODE REVIEW

**Comment by javu7762** [comment]
@brb 
I did it.

**Comment by javu7762** [comment]
@aanm 
I fixed CI failure.

File: pkg/service/service.go
Comment (aanm) [inline pkg/service/service.go]: I'm not sure if this is what we want here. the `hash` on its own prevents on having duplicated backends. That's why we are storing it by hash.

File: pkg/service/service_test.go
Comment (aanm) [inline pkg/service/service_test.go]: please add this test:
```
	different = []*lb.Backend{
		lb.NewBackend(0, lb.UDP, net.ParseIP("10.0.0.5"), 8080),
		lb.NewBackend(0, lb.TCP, net.ParseIP("10.0.0.5"), 8080),
	}
```
this should not fail but my understanding is that the tests will fail for this case.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/service/service_test.go
Comment (javu7762) [reply pkg/service/service_test.go]: @aanm 
I added  that testcase with adding logic for checking protocols

File: pkg/service/service.go
Comment (javu7762) [reply pkg/service/service.go]: As you know, `hash` does not care protocols.
If you have better idea, let me know and  I will appreciate.

**Review by aanm — APPROVED** [review]

**Review by nbusseneau — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Review by brb — CHANGES_REQUESTED** [review]
LGTM, just could you squash all commits into a single one?

**Comment by javu7762** [comment]
@brb 
I squashed all  into one commit.

**Review by brb — APPROVED** [review]
Thanks!

**Comment by brb** [comment]
No need to re-run the tests, marking as ready to merge.

**Review by rolinh — APPROVED** [review]

**Comment by brb** [comment]
/test-1.25-net-next

File: pkg/testutils/mockmaps/lbmap.go
Comment (aditighag) [inline pkg/testutils/mockmaps/lbmap.go:66]: Why is this mock maps code getting deleted? Is this intentional?

Comment (javu7762) [reply pkg/testutils/mockmaps/lbmap.go:66]: @aditighag 
I deleted mock code intentionally because when backend count  changes by assigning backend array in service testcase,   test case fails.

**Comment by javu7762** [comment]
@brb 
Is there anything to do by me for merging?

**Comment by brb** [comment]
>  Is there anything to do by me for merging?

@javu7762 There is a conflict in ` pkg/service/service_test.go`. Also, we need to get ACK from @aditighag.

File: pkg/testutils/mockmaps/lbmap.go
Comment (javu7762) [reply pkg/testutils/mockmaps/lbmap.go:66]: @aditighag 
Is there problem in deleting mock code?

**Comment by javu7762** [comment]
@brb 
Thank you. 
I will do it.

**Comment by aditighag** [comment]
https://github.com/cilium/cilium/pull/20883#discussion_r964621800

>As you know, hash does not care protocols.

The hash does account for L4 protocols, but we currently ignore it because it's not supported - https://github.com/cilium/cilium/blob/master/pkg/loadbalancer/loadbalancer.go#L722-L722. Can we instead fix the `Hash` function to account for the protocol field? We've been ignoring the protocol field until now though. I'll let @brb chime in here since he authored the bpf mock map code the PR is deleting.

Either way, my comments are not blocking as you already have all the approvals. Please drop the merge commits, and rebase your PR.

**Comment by javu7762** [comment]
@aditighag 

Thank you for reply.

> Can we instead fix the Hash function to account for the protocol field? 

I think It makes big changes including bpf maps and migration.

>Either way, my comments are not blocking as you already have all the approvals. Please drop the merge commits, and rebase your PR.

I did it.

@brb 
Is there anything to do?  
If you let me know, I will thank you

**Comment by javu7762** [comment]
@brb 
I think this PR is ready to merge.
Can you check it out?

**Review by brb — COMMENTED** [review]
I actually have one reservation regarding this change. Maybe someone is actually using same backends to implement a poor man's weighted LB?

**Comment by javu7762** [comment]
@brb 
If someone will implement new feature which is supporting LB with same backends, this PR is useless. 
so  you think this PR has to be closed. Right?

I will wait for your reply.

**Comment by brb** [comment]
>  so you think this PR has to be closed. Right?

Yep, I suggest so.

**Comment by brb** [comment]
>  so you think this PR has to be closed. Right?

I missed that the deletion was not working in that case. So, it's very unlikely that someone was using this hack to implement the weights.

**Pull request closed without merging**
